src/mgmt/srv_pool.c

```diff
@@ -659,16 +678,65 @@ ds_mgmt_pool_list(void)
 
 	D_DEBUG(DB_MGMT, "pools:\n");
 	rc = rdb_tx_iterate(&tx, &svc->ms_pools, false /* !backward */,
-			    enum_pool_cb, NULL /* arg */);
+			    enum_pool_cb, &iter_args);
 
 	ABT_rwlock_unlock(svc->ms_lock);
 	rdb_tx_end(&tx);
+
 out_svc:
 	ds_mgmt_svc_put_leader(svc);
 out:
+	if (rc == 0) {
+		*npools = iter_args.npools;
+		*poolsp = iter_args.pools;
+		*pools_len = iter_args.pools_index;
+	} else {
+		if (iter_args.pools) {
+			int pc;
+
+			for (pc = 0; pc < iter_args.pools_index; pc++)
+				d_rank_list_free(iter_args.pools[pc].lp_svc);
+			D_FREE(iter_args.pools);
+		}
+	}
+
 	return rc;
 }
 
+/* TODO: dRPC handler for mgmt service list pools (daos_shell / golang) */
+
+/* CaRT RPC handler for management service "list pools" (dmg / C API) */
+void
+ds_mgmt_hdlr_list_pools(crt_rpc_t *rpc_req)
+{
+	struct mgmt_list_pools_in	*pc_in;
+	struct mgmt_list_pools_out	*pc_out;
+	struct mgmt_list_pools_one	*pools = NULL;
+	size_t				 pools_len = 0;
+	uint64_t			 npools;
+	int				 rc;
+
+	pc_in = crt_req_get(rpc_req);
+	D_ASSERT(pc_in != NULL);
+	pc_out = crt_reply_get(rpc_req);
+	D_ASSERT(pc_out != NULL);
+
+	npools = pc_in->lp_npools;
+
+	rc = ds_mgmt_list_pools(pc_in->lp_grp, &npools, &pools,
+				&pools_len);
+
+	pc_out->lp_npools = npools;
+	pc_out->lp_rc = rc;
+	pc_out->lp_pools.ca_arrays = pools;
+	pc_out->lp_pools.ca_count = pools_len;
+
+	/* TODO: something different for larger RPC replies? */
+	rc = crt_reply_send(rpc_req);
+	if (rc != 0)
+		D_ERROR("crt_reply_send failed, rc: %d\n", rc);
+}
```

Comment:
Do we need to release pools (the variable)?

Reply:
Yes, will do.